File: app/src/main/java/com/example/jebeniis/MainActivity.java
package com.example.jebeniis;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;

public final class MainActivity extends AppCompatActivity {
    RecyclerView recyclerView;
    Adapter adapter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        HashMap<String, String> questions = getQuestions();

        adapter = new Adapter(this, new ArrayList<String>(questions.keySet()), questions);

        recyclerView = findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(adapter);
        adapter.notifyDataSetChanged();
    }

    public HashMap<String, String> getQuestions() {
        String content = loadAsset("is.json");
        HashMap<String, String> questions = new HashMap<>();
        try {
            JSONObject jsonObject = new JSONObject(content);
            jsonObject.keys().forEachRemaining(key -> {
                try {
                    String value = jsonObject.getString(key);
                    questions.put(key, value);
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            });
            return questions;
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    public String loadAsset(String asset) {
        String json = null;
        try {
            InputStream is = this.getAssets().open(asset);
            int size = is.available();
            byte[] buffer = new byte[size];
            is.read(buffer);
            is.close();
            json = new String(buffer, "UTF-8");
        } catch (IOException ex) {
            ex.printStackTrace();
            return null;
        }
        return json;
    }
}

File: app/src/main/java/com/example/jebeniis/MainActivity.kt
package com.example.jebeniis

import android.os.Bundle
import android.view.KeyEvent
import android.widget.EditText
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.apache.commons.lang3.StringUtils
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.util.*

class MainActivity : AppCompatActivity() {
    private lateinit var recyclerView: RecyclerView
    private var adapter: Adapter? = null
    private lateinit var searchBox: EditText
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val questions = questions
        val keyList = ArrayList(
            questions!!.keys
        )
        searchBox = findViewById(R.id.searchBox)
        searchBox.setOnEditorActionListener{ _: TextView?, _: Int, _: KeyEvent? ->
            val search = searchBox.text.toString()
            val filteredKeys: MutableList<FilterItem> = ArrayList()
            for (key in keyList) {
                key.lowercase(Locale.getDefault())
                var minLev = 10000
                val words = key.split(" ").toTypedArray()
                for (word in words) {
                    @Suppress("DEPRECATION") val lev = StringUtils.getLevenshteinDistance(word, search)
                    if (lev < minLev) {
                        minLev = lev
                    }
                }
                if (minLev < 8) {
                    filteredKeys.add(FilterItem(key, minLev))
                }
            }
            filteredKeys.sortWith{ o1: FilterItem, o2: FilterItem -> o1.lev - o2.lev }
            keyList.removeIf { true }
            for (item in filteredKeys) {
                keyList.add(item.key)
            }
            true
        }
        adapter = Adapter(this, keyList, questions)
        recyclerView = findViewById(R.id.recyclerView)
        recyclerView.layoutManager = LinearLayoutManager(this)
        adapter.also { recyclerView.adapter = it }
    }

    private val questions: HashMap<String, String>?
        get() {
            val content = "is.json".loadAsset()
            val questions = HashMap<String, String>()
            try {
                val jsonObject = content?.let { JSONObject(it) }
                jsonObject?.keys()?.forEachRemaining { key: String ->
                    try {
                        val value = jsonObject.getString(key)
                        key.lowercase(Locale.getDefault())
                        questions[key] = value
                    } catch (e: JSONException) {
                        e.printStackTrace()
                    }
                }
                return questions
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            return null
        }

    private fun String.loadAsset(): String? {
        val json: String = try {
            val `is` = this@MainActivity.assets.open(this)
            val size = `is`.available()
            val buffer = ByteArray(size)
            `is`.read(buffer)
            `is`.close()
            String(buffer, StandardCharsets.UTF_8)
        } catch (ex: IOException) {
            ex.printStackTrace()
            return null
        }
        return json
    }
}